File: scraper.py
import os
import yt_dlp
import time
import threading
import logging
from telegram import InlineQueryResultArticle, Update, InputFile, InlineQueryResultCachedVideo, InlineQueryResultCachedPhoto, InputTextMessageContent
from telegram.ext import Updater, CommandHandler, MessageHandler, Filters, InlineQueryHandler
from collections import defaultdict
from datetime import datetime, timedelta

# Configure logging
logging.basicConfig(
    level=logging.DEBUG,
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s',
    handlers=[
        logging.StreamHandler(),
        logging.FileHandler('scraper.log')
    ]
)
logger = logging.getLogger(__name__)

# Load secrets from JSON file
try:
    import json
    with open('secrets.json') as f:
        secrets = json.load(f)
    BOT_TOKEN = secrets['bot_token']
    CHAT_ID = secrets.get('chat_id', 824640741)  # Default fallback
except Exception as e:
    logger.error(f"Failed to load secrets: {e}")
    raise RuntimeError("Missing required secrets configuration")
DOWNLOADS_DIR = os.path.join(os.path.dirname(__file__), "downloads")
os.makedirs(DOWNLOADS_DIR, exist_ok=True)
CLEANUP_INTERVAL = 30
REQUEST_LIMIT = 15
MAX_CONCURRENT_DOWNLOADS = 30

user_requests = defaultdict(list)
lock = threading.Lock()

def download_instagram_media(url: str, user_id: str):
    """Reliable Instagram media downloader using yt-dlp"""
    logger.info(f"Starting download for {url}")
    user_dir = os.path.join(DOWNLOADS_DIR, str(user_id))
    os.makedirs(user_dir, exist_ok=True)
    timestamp = str(int(time.time()))
    file_path = os.path.join(user_dir, f'{timestamp}_instagram_media')

    ydl_opts = {
        'outtmpl': os.path.join(user_dir, f'{timestamp}_%(title)s.%(ext)s'),
        'quiet': True,
        'socket_timeout': 15,
        'retries': 3,
        'format': 'best',
        'writethumbnail': False,
        'nocheckcertificate': True,
        'extractor_args': {
            'instagram': {
                'extract_flat': True
            }
        }
    }

    try:
        with yt_dlp.YoutubeDL(ydl_opts) as ydl:
            info = ydl.extract_info(url, download=True)
            file_path = ydl.prepare_filename(info)
            
            if not os.path.exists(file_path):
                raise Exception("Download failed - file not created")
            
            is_video = file_path.lower().endswith(('.mp4', '.mkv', '.webm'))
            
            return {
                'path': file_path,
                'is_video': is_video,
                'title': info.get('title', 'Instagram Media')[:64],
                'user_id': user_id
            }
    except Exception as e:
        logger.error(f"Download failed: {str(e)}")
        return None

def check_rate_limit(user_id):
    """Implement rate limiting per user"""
    with lock:
        now = datetime.now()
        user_requests[user_id] = [t for t in user_requests[user_id] if now - t < timedelta(minutes=1)]
        if len(user_requests[user_id]) >= REQUEST_LIMIT:
            return False
        user_requests[user_id].append(now)
        return True

def clean_old_files():
    """Remove files older than CLEANUP_INTERVAL"""
    now = time.time()
    for user_dir in os.listdir(DOWNLOADS_DIR):
        user_path = os.path.join(DOWNLOADS_DIR, user_dir)
        if os.path.isdir(user_path):
            for filename in os.listdir(user_path):
                filepath = os.path.join(user_path, filename)
                if os.path.isfile(filepath):
                    file_time = os.path.getmtime(filepath)
                    if now - file_time > CLEANUP_INTERVAL:
                        try:
                            os.remove(filepath)
                        except Exception as e:
                            logger.error(f"Error cleaning file {filepath}: {e}")

def start(update: Update, context):
    update.message.reply_text("Send me an Instagram/Youtube link, and I'll fetch the media for you!")

def handle_message(update: Update, context):
    user_id = update.message.from_user.id
    url = update.message.text.strip()
    
    if not url.startswith(('http://', 'https://')):
        update.message.reply_text("Please send a valid URL starting with http:// or https://")
        return
    
    if not check_rate_limit(user_id):
        update.message.reply_text("⚠️ Please wait a minute before making more requests")
        return
    
    update.message.reply_text("⏳ Downloading media, please wait...")
    
    result = download_instagram_media(url, user_id)
    if result:
        try:
            with open(result['path'], "rb") as media_file:
                if result['is_video']:
                    update.message.reply_video(
                        video=InputFile(media_file),
                        caption="🎥 Here's your video!"
                    )
                else:
                    update.message.reply_photo(
                        photo=InputFile(media_file),
                        caption="📸 Here's your photo!"
                    )
        except Exception as e:
            update.message.reply_text("❌ Failed to send media. Please try again.")
            logger.error(f"Error sending media to {user_id}", exc_info=True)
    else:
        update.message.reply_text("❌ Failed to download media. Please check the URL and try again.")

def inline_query(update: Update, context):
    user_id = update.inline_query.from_user.id
    query = update.inline_query.query.strip()
    
    if not query:
        return
    
    if not query.startswith(('http://', 'https://')):
        update.inline_query.answer([
            InlineQueryResultArticle(
                id="invalid_url",
                title="Invalid URL",
                input_message_content=InputTextMessageContent("Please provide a valid Instagram URL starting with http:// or https://")
            )
        ])
        return
    
    if not check_rate_limit(user_id):
        update.inline_query.answer([
            InlineQueryResultArticle(
                id="rate_limit",
                title="Too many requests",
                input_message_content=InputTextMessageContent("⚠️ Please wait a minute before making more requests")
            )
        ])
        return
    
    clean_old_files()
    
    result = download_instagram_media(query, user_id)
    if not result:
        update.inline_query.answer([
            InlineQueryResultArticle(
                id="error",
                title="Download failed",
                input_message_content=InputTextMessageContent("❌ Failed to download media. Please try a different link.")
            )
        ])
        return
    
    try:
        with open(result['path'], 'rb') as media_file:
            if result['is_video']:
                video_msg = context.bot.send_video(
                    chat_id=CHAT_ID,
                    video=InputFile(media_file),
                    disable_notification=True
                )
                update.inline_query.answer([
                    InlineQueryResultCachedVideo(
                        id="1",
                        title=result['title'],
                        video_file_id=video_msg.video.file_id,
                        caption="🎥 Here's your video!"
                    )
                ])
            else:
                photo_msg = context.bot.send_photo(
                    chat_id=CHAT_ID,
                    photo=InputFile(media_file),
                    disable_notification=True
                )
                update.inline_query.answer([
                    InlineQueryResultCachedPhoto(
                        id="1",
                        title=result['title'],
                        photo_file_id=photo_msg.photo[0].file_id,
                        caption="📸 Here's your photo!"
                    )
                ])
    except Exception as e:
        logger.error("Error processing inline query", exc_info=True)
        update.inline_query.answer([
            InlineQueryResultArticle(
                id="error",
                title="Processing failed",
                input_message_content=InputTextMessageContent("❌ Failed to process media. Please try again.")
            )
        ])

def scheduled_cleanup():
    """Periodically clean up old files"""
    while True:
        time.sleep(CLEANUP_INTERVAL)
        clean_old_files()

def main():
    cleanup_thread = threading.Thread(target=scheduled_cleanup, daemon=True)
    cleanup_thread.start()
    
    updater = Updater(BOT_TOKEN, use_context=True)
    dp = updater.dispatcher
    
    dp.add_handler(CommandHandler("start", start))
    dp.add_handler(MessageHandler(Filters.text & ~Filters.command, handle_message))
    dp.add_handler(InlineQueryHandler(inline_query))
    
    logger.info("Bot is running with multi-user support...")
    updater.start_polling()
    updater.idle()

if __name__ == "__main__":
    main()

File: scraper.go
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"log"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"time"

	tgbotapi "github.com/go-telegram-bot-api/telegram-bot-api/v5"
)

type Config struct {
	BotToken string `json:"bot_token"`
	ChatID   int64  `json:"chat_id"`
}

var (
	config      Config
	downloadDir = "downloads"
)

func main() {
	// Load config
	loadConfig("secrets.json")

	// Create bot
	bot, err := tgbotapi.NewBotAPI(config.BotToken)
	if err != nil {
		log.Panic(err)
	}

	bot.Debug = true
	log.Printf("Authorized on account %s", bot.Self.UserName)

	// Set up update config with offset
	u := tgbotapi.NewUpdate(0)
	u.Timeout = 60
	u.Offset = 0 // Explicitly set offset

	// Get initial updates to clear the queue
	if _, err := bot.GetUpdates(u); err != nil {
		log.Printf("Warning: failed to clear update queue: %v", err)
	}

	// Create buffered updates channel
	updates := bot.GetUpdatesChan(u)

	// Handle updates with recovery
	for {
		select {
		case update, ok := <-updates:
			if !ok {
				log.Println("Updates channel closed, reconnecting...")
				time.Sleep(5 * time.Second)
				updates = bot.GetUpdatesChan(u)
				continue
			}

			if update.Message != nil {
				handleMessage(bot, update.Message)
			} else if update.InlineQuery != nil {
				handleInlineQuery(bot, update.InlineQuery)
			}
		}
	}
}

func loadConfig(filename string) {
	file, err := os.ReadFile(filename)
	if err != nil {
		log.Panic(err)
	}

	err = json.Unmarshal(file, &config)
	if err != nil {
		log.Panic(err)
	}
}

func handleMessage(bot *tgbotapi.BotAPI, msg *tgbotapi.Message) {
	if !strings.HasPrefix(msg.Text, "http") {
		reply := tgbotapi.NewMessage(msg.Chat.ID, "Please send a valid URL starting with http:// or https://")
		bot.Send(reply)
		return
	}

	// Send processing message
	processingMsg := tgbotapi.NewMessage(msg.Chat.ID, "⏳ Downloading media, please wait...")
	bot.Send(processingMsg)

	// Download media
	result, err := downloadMedia(msg.Text, msg.From.ID)
	if err != nil {
		errorMsg := tgbotapi.NewMessage(msg.Chat.ID, "❌ Failed to download media. Please check the URL and try again.")
		bot.Send(errorMsg)
		return
	}

	// Send media
	if result.IsVideo {
		video := tgbotapi.NewVideo(msg.Chat.ID, tgbotapi.FilePath(result.Path))
		bot.Send(video)
	} else {
		photo := tgbotapi.NewPhoto(msg.Chat.ID, tgbotapi.FilePath(result.Path))
		bot.Send(photo)
	}
}

func handleInlineQuery(bot *tgbotapi.BotAPI, query *tgbotapi.InlineQuery) {
	if !strings.HasPrefix(query.Query, "http") {
		return
	}

	// Download media with debug logging
	log.Printf("Processing inline query for URL: %s", query.Query)
	result, err := downloadMedia(query.Query, query.From.ID)
	if err != nil {
		log.Printf("Download failed for %s: %v", query.Query, err)
		return
	}
	log.Printf("Downloaded media: %s (isVideo: %v)", result.Path, result.IsVideo)

	// Validate chat ID exists
	if config.ChatID == 0 {
		log.Printf("Invalid chat ID configured")
		return
	}

	// Send media to chat to get file ID
	var fileID string
	if result.IsVideo {
		video := tgbotapi.NewVideo(config.ChatID, tgbotapi.FilePath(result.Path))
		msg, err := bot.Send(video)
		if err != nil {
			return
		}
		fileID = msg.Video.FileID
	} else {
		photo := tgbotapi.NewPhoto(config.ChatID, tgbotapi.FilePath(result.Path))
		msg, err := bot.Send(photo)
		if err != nil {
			log.Printf("Failed to send photo: %v", err)
			return
		}
		if len(msg.Photo) == 0 {
			log.Printf("No photo data in message response")
			return
		}
		fileID = msg.Photo[0].FileID
		log.Printf("Got photo file ID: %s", fileID)
	}

	// Answer inline query
	var results []interface{}
	if result.IsVideo {
		// Extract filename for title
		_, filename := filepath.Split(result.Path)
		videoResult := tgbotapi.NewInlineQueryResultCachedVideo("1", fileID, filename)
		results = append(results, videoResult)
	} else {
		photoResult := tgbotapi.NewInlineQueryResultCachedPhoto("1", fileID)
		results = append(results, photoResult)
	}

	inlineConfig := tgbotapi.InlineConfig{
		InlineQueryID: query.ID,
		Results:       results,
		CacheTime:     0,
	}

	bot.Send(inlineConfig)
}

type DownloadResult struct {
	Path    string
	IsVideo bool
}

func downloadMedia(url string, userID int64) (*DownloadResult, error) {
	// Create user directory
	userDir := filepath.Join(downloadDir, fmt.Sprintf("%d", userID))
	os.MkdirAll(userDir, os.ModePerm)
	timestamp := time.Now().Unix()

	// Check if URL is Instagram
	if strings.Contains(url, "instagram.com") {
		// Use yt-dlp for Instagram downloads
		outputTemplate := filepath.Join(userDir, fmt.Sprintf("%d_%%(title)s.%%(ext)s", timestamp))
		cmd := exec.Command("yt-dlp",
			"-o", outputTemplate,
			"--no-playlist",
			"--force-overwrites",
			url,
		)

		var out bytes.Buffer
		var stderr bytes.Buffer
		cmd.Stdout = &out
		cmd.Stderr = &stderr

		err := cmd.Run()
		if err != nil {
			if strings.Contains(stderr.String(), "Private") || strings.Contains(stderr.String(), "Restricted") {
				return nil, fmt.Errorf("this bot doesn't work with private or restricted Instagram posts")
			}
			return nil, fmt.Errorf("failed to download Instagram media: %v\n%s", err, stderr.String())
		}

		// Find downloaded file
		matches, _ := filepath.Glob(filepath.Join(userDir, fmt.Sprintf("%d_*", timestamp)))
		if len(matches) == 0 {
			return nil, fmt.Errorf("no file downloaded from Instagram")
		}

		filePath := matches[0]
		return &DownloadResult{
			Path:    filePath,
			IsVideo: strings.HasSuffix(strings.ToLower(filePath), ".mp4"),
		}, nil
	}

	// Fall back to yt-dlp for non-Instagram URLs
	outputTemplate := filepath.Join(userDir, fmt.Sprintf("%d_%%(title)s.%%(ext)s", timestamp))

	// Try with cookies first if available
	args := []string{
		"-o", outputTemplate,
		"--no-playlist",
		"--extractor-args", "instagram:include_dash_manifest=true;flat=1",
		"--force-overwrites",
		"--retries", "3",
		"--format", "(bestvideo+bestaudio/best)[protocol^=http]",
		"--merge-output-format", "mp4",
		"--convert-thumbnails", "jpg",
		"--verbose",
		"--add-header", "User-Agent:Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36",
		"--add-header", "Referer:https://www.instagram.com/",
		"--add-header", "Origin:https://www.instagram.com",
	}

	// Only use cookies.txt if exists
	if _, err := os.Stat("cookies.txt"); err == nil {
		args = append(args, "--cookies", "cookies.txt")
	} else {
		log.Printf("Note: cookies.txt not found - some Instagram content may require login")
	}

	args = append(args, url)
	cmd := exec.Command("yt-dlp", args...)

	var stdout, stderr bytes.Buffer
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr

	err := cmd.Run()
	if err != nil {
		fullError := fmt.Sprintf("yt-dlp error: %v\n=== Stdout ===\n%s\n=== Stderr ===\n%s",
			err, stdout.String(), stderr.String())
		log.Printf(fullError)
		return nil, fmt.Errorf(fullError)
	}

	// Get downloaded file
	matches, _ := filepath.Glob(filepath.Join(userDir, fmt.Sprintf("%d_*", timestamp)))
	if len(matches) == 0 {
		return nil, fmt.Errorf("no file downloaded")
	}

	filePath := matches[0]
	isVideo := strings.HasSuffix(strings.ToLower(filePath), ".mp4") ||
		strings.HasSuffix(strings.ToLower(filePath), ".mkv") ||
		strings.HasSuffix(strings.ToLower(filePath), ".webm") ||
		strings.HasSuffix(strings.ToLower(filePath), ".mov")

	return &DownloadResult{
		Path:    filePath,
		IsVideo: isVideo,
	}, nil
}
